### PR TITLE
feat: implement search on mentor registry page

### PR DESCRIPTION
**Pull request by Alphajax**
**🟢 Add `deploy` label if you want to deploy this Pull Request to staging environment**

#### 🧑‍⚖️ Pull Request Naming Convention

- Title should follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary)
- Do not put issue id in title
- Do not put WIP in title. Use [Draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/) functionality
- Consider to add `area:*` label(s)

* [x] I followed naming convention rules

---

#### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Performance optimization
- [ ] Refactoring
- [ ] Test Case
- [ ] Documentation update
- [ ] Other

#### 🔗 Related issue link

 [related issue link](https://github.com/rolling-scopes/rsschool-app/issues/2140)

#### 💡 Background and solution

- implement filtering mentor registry on backend

#### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Database migration is added or not needed
- [x] Documentation is updated/provided or not needed
- [x] Changes are tested locally

## Summary by CodeRabbit

- **New Features**
	- Introduced advanced filtering options for mentor registries, including by GitHub ID and city name.
	- Implemented pagination for the mentor registry to enhance usability and performance.
	- Added new response interfaces to better structure the data for mentor registries.

- **Enhancements**
	- Updated mentor registry API to support new filtering and pagination capabilities.
	- Enhanced mentor registry table with pagination controls and updated filtering logic.

- **Documentation**
	- Updated API documentation to reflect new query parameters and response structures for mentor registry operations.

- **Refactor**
	- Refactored mentor registry service to include a new method for filtered and paginated results.
	- Improved code structure for handling mentor registry data.

### CODE REVIEW

File: nestjs/src/registry/registry.service.ts [hunk 1]
```diff
@@ -85,4 +86,71 @@ export class RegistryService {
...
+  }: {
+    githubId?: string;
+    cityName?: string;
+    page: number;
+    limit: number;
+    preselectedCourses?: number[];
+    preferedCourses?: number[];
+    technicalMentoring?: string[];
+  }) {
+    console.log('here');
```
Comment (valerydluski) [inline nestjs/src/registry/registry.service.ts:107]: 😃

File: client/src/modules/MentorRegistry/components/MentorRegistryTableContainer.tsx [hunk 2]
```diff
@@ -337,17 +384,51 @@ export const MentorRegistryTableContainer = ({
   };
 
   const handleClearAllButtonClick = () => {
-    setCombinedFilter({ preferredCourses: [], technicalMentoring: [], preselectedCourses: [] });
+    setCombinedFilter({
+      preferredCourses: [],
+      technicalMentoring: [],
+      preselectedCourses: [],
+      githubId: [],
+      cityName: [],
```
Comment (valerydluski) [inline client/src/modules/MentorRegistry/components/MentorRegistryTableContainer.tsx:392]:
```suggestion
    preferredCourses: string[];
    technicalMentoring: string[];
    preselectedCourses: string[];
    githubId: string[];
    cityName: string[];
```

File: client/src/modules/MentorRegistry/components/MentorRegistryTableContainer.tsx [hunk 3]
```diff
@@ -337,17 +384,51 @@ export const MentorRegistryTableContainer = ({
...
     setTagFilters([]);
   };
 
+  useAsync(async () => {
+    try {
+      const { content, total } = await mentorRegistryService.filterMentorRegistries({
+        pageSize: PAGINATION,
+        currentPage,
+        githubId: combinedFilter.githubId?.length ? (combinedFilter.githubId[0] as string) : undefined,
+        cityName: combinedFilter.cityName?.length ? (combinedFilter.cityName[0] as string) : undefined,
```
Comment (valerydluski) [inline client/src/modules/MentorRegistry/components/MentorRegistryTableContainer.tsx:403]: Can something like this be used?
`combinedFilter.githubId?.[0] ?? undefined` or `combinedFilter.githubId?.[0] ?? undefined`

**Comment by valerydluski** [comment]
Thank you!

File: client/src/modules/MentorRegistry/components/MentorRegistryTableContainer.tsx (hunk 2 above)
Comment (Alphajax) [reply client/src/modules/MentorRegistry/components/MentorRegistryTableContainer.tsx:392]: looks strange as it's passing props to function but not type declaration

Comment (valerydluski) [reply client/src/modules/MentorRegistry/components/MentorRegistryTableContainer.tsx:392]: my bad

File: client/src/modules/MentorRegistry/components/MentorRegistryTableContainer.tsx [hunk 4]
```diff
@@ -329,6 +375,7 @@ export const MentorRegistryTableContainer = ({
           }));
         }
         break;
+      //todo add here
```
[the remark this answers is not in the document]
Comment (Alphajax) [reply client/src/modules/MentorRegistry/components/MentorRegistryTableContainer.tsx]: no need, deleted

File: client/src/services/mentorRegistry.ts [hunk 5]
```diff
@@ -17,6 +17,31 @@ export interface MentorRegistry {
...
+
+export interface GetMentorRegistriesOptions {
+  currentPage?: number;
+  pageSize?: number;
+  githubId?: string;
+  cityName?: string;
+  preferedCourses?: number[];
+  preselectedCourses?: number[];
+  technicalMentoring?: string[];
+}
```
Comment (valerydluski) [inline client/src/services/mentorRegistry.ts:43]: DTO Adoption from Backend: Assuming you can directly use or extend DTOs defined by your backend, align your frontend interfaces with these DTOs. This approach reduces duplication and mismatches between the frontend and backend.
```
export interface GetMentorRegistriesOptions extends Partial<GetMentorRegistriesDto> {
}
```

**Review by valerydluski — APPROVED** [review]

File: nestjs/src/registry/registry.service.ts [hunk 6]
```diff
@@ -85,4 +86,70 @@ export class RegistryService {
...
+    technicalMentoring?: string[];
+  }) {
+    const req = this.getPreparedMentorRegistriesQuery();
+    if (githubId) {
+      req.andWhere(`"user"."githubId" ILIKE :githubId`, { githubId: `%${githubId}%` });
+    }
+    if (cityName) {
+      req.andWhere(`"user"."cityName" ILIKE :cityName`, { cityName: `%${cityName}%` });
+    }
+    if (preselectedCourses && preselectedCourses.length) {
```
Comment (valerydluski) [inline nestjs/src/registry/registry.service.ts:114]:
```suggestion
    if (preselectedCourses?.length) {
```

File: nestjs/src/registry/registry.service.ts [hunk 7]
```diff
@@ -85,4 +86,70 @@ export class RegistryService {
...
+      req.andWhere(
+        `EXISTS (
+          SELECT
+          FROM unnest(string_to_array(mentorRegistry.preselectedCourses, ',')) course
+          WHERE course = ANY(:preselectedCourses)
+        )`,
+        { preselectedCourses },
+      );
+    }
+    if (preferedCourses && preferedCourses.length) {
```
Comment (valerydluski) [inline nestjs/src/registry/registry.service.ts:124]:
```suggestion
    if (preferedCourses?.length) {
```

File: nestjs/src/registry/registry.service.ts [hunk 8]
```diff
@@ -85,4 +86,70 @@ export class RegistryService {
...
+      req.andWhere(
+        `EXISTS (
+        SELECT
+        FROM unnest(string_to_array(mentorRegistry.preferedCourses, ',')) course
+        WHERE course = ANY(:preferedCourses)
+      )`,
+        { preferedCourses },
+      );
+    }
+    if (technicalMentoring && technicalMentoring.length) {
```
Comment (valerydluski) [inline nestjs/src/registry/registry.service.ts:134]:
```suggestion
    if (technicalMentoring?.length) {
```

**Pull request merged**